### PR TITLE
[b/339392173] implement HdfsPermissionExtractionConnector

### PR DESCRIPTION
**Pull request by sten1ee**

### CODE REVIEW

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/AtomicCounter.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+class AtomicCounter {
+
+  private int currentCount;
+
+  AtomicCounter(int currentCount) {
+    this.currentCount = currentCount;
+  }
+
+  synchronized void increment() {
+    assert currentCount < Integer.MAX_VALUE : "Implementation limits reached!";
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/AtomicCounter.java:32]: Please change `assert` to `checkArgument` call. Possibly in a separate PR.

**Review by dawidxc — APPROVED** [review]

**Review by shevek-google — COMMENTED** [review]
This PR should probably use `try(ExecutorManager) { }`

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/io/OutputHandle.java [hunk 2]
```diff
@@ -48,8 +49,14 @@ public default CharSink asCharSink(@Nonnull Charset charset) throws IOException
   /**
    * Renames the temporary file to the final file.
    *
-   * <p>The stream, if any, must be closed. The temporary file must exist. Do NOT call commit() in a
-   * try-with-resources or finally block, or you risk committing bad data.
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/io/OutputHandle.java:52]: What invalidated this comment?

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/SingleDirScanJob.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  protected Void compute() throws IllegalStateException {
+    long numFiles = 0;
+    long numDirs = 0;
+    long accumFileSize = 0;
+
+    try {
+      scanCtx.beginWalkDir(dir);
+      for (FileStatus file : scanCtx.listDirectory(dir)) {
+        // Process file or dir (in this case - just collect statistics)
+        accumFileSize += file.getLen();
```
Comment (paolomorandini) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/SingleDirScanJob.java:48]: What does `getLen` return for directories?

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/SingleDirScanJob.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        if (file.isDirectory()) {
+          numDirs++;
+          new SingleDirScanJob(scanCtx, jobsTodoCounter, file).fork();
+        } else {
+          scanCtx.walkFile(file);
+        }
+      }
+      scanCtx.endWalkDir(dir, numFiles, numDirs, accumFileSize);
+    } catch (Exception exn) {
+      LOG.error(
```
Comment (paolomorandini) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/SingleDirScanJob.java:60]: The dumper conventionally saves also exceptions (with stack trace) in the produced artifact.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/io/OutputHandle.java (hunk 2 above)
Comment (sten1ee) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/io/OutputHandle.java:52]: Dawid, did some research and concluded it is safe to delete, IIRC.
@dawidxc am I retelling my fake memories?

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/ScanContext.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+      csvPrinter.printRecord(
+          dir.getPath().toUri().getPath(),
+          dir.getOwner(),
+          dir.getGroup(),
+          dir.getPermission().toString());
+    }
+  }
+
+  void walkFile(FileStatus file) {}
```
[the remark this answers is not in the document]
Comment (sten1ee) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/ScanContext.java:90]: Yes, but this one is being invoked, so it is a placeholder.
I am sure the VM will optimize the call away.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/ScanContext.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    FileStatus[] files = fs.listStatus(dir.getPath());
+    synchronized (LOCK1) {
+      timeSpentInListStatus =
+          timeSpentInListStatus.plus(Duration.between(instantListBegin, Instant.now()));
+      numFilesByListStatus += files.length;
+    }
+    return files;
+  }
+
+  void beginWalkDir(FileStatus dir) {}
```
[the remark this answers is not in the document]
Comment (sten1ee) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/ScanContext.java:73]: Yes, but this one is being invoked, so it is a placeholder.
I am sure the VM will optimize the call away.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/SingleDirScanJob.java (hunk 3 above)
Comment (sten1ee) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/SingleDirScanJob.java:48]: According to the JavaDoc of that method:
```
Get the length of this file, in bytes.
Returns:
the length of this file, in bytes.
```

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/HdfsPermissionExtractionConnector.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+@Description("Dumps permissions from the HDFS.")
+public class HdfsPermissionExtractionConnector extends AbstractConnector {
+
+  public HdfsPermissionExtractionConnector() {
+    super("hdfs-permissions");
+  }
+
+  @Nonnull
+  @Override
+  public String getDefaultFileName(boolean isAssessment, Clock clock) {
```
[the remark this answers is not in the document]
Comment (misolt) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/HdfsPermissionExtractionConnector.java:57]: `getDefaultFileName` is called from MetadataDumper, which currently doesn't know if a given connector is logs or metadata or something else. We get all services that implement Connector and call some Connector methods on them.

If we want to remove this method and avoid casting such as `LogsConnector lc = (LogsConnector) connector`, then I see the 2 options:
- have separate code in MetadataDumper to get and handle different categories of connectors
- move the 2 parameters to connector attributes, probably adding a new class to handle this logic for Connector

but both of these options would probably lead to more code for solving a relatively simple problem

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/HdfsPermissionExtractionTask.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class HdfsPermissionExtractionTask implements Task<Void> {
```
[the remark this answers is not in the document]
Comment (sten1ee) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/HdfsPermissionExtractionTask.java:37]: fixed with https://github.com/google/dwh-migration-tools/pull/479

**Pull request merged**